<a href="https://colab.research.google.com/github/santhoshbvsrk/CapstoneProjects/blob/main/DataAnalytics_CDF/InsaidTelecom_CDFCapstone.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 

*  As the dataset is huge, we'll zip it, upload the zip file to google drive and mount your google drive to colab, unzip it to access data from your colab.


*   Import the drive package from google.colab library to mount your google drive on to google colab.





In [2]:
from google.colab import drive
drive.mount ('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).




*   Once mounting is completed, mention the path of your zip file which contains the data to read.

In [3]:
!unzip gdrive/My\ Drive/CDF_Capstone_Project/events_data.zip

Archive:  gdrive/My Drive/CDF_Capstone_Project/events_data.zip
  inflating: events_data.csv         




*   Once mounting and unzipping is completed successfully, you should be able to see the zip file contains in your colab folder.


In [4]:
import pandas as pd 
import numpy as np

In [5]:
events_data_df=pd.read_csv("events_data.csv")
events_data_df.head()

event_id     device_id            timestamp  ...   latitude      city       state
0   2765368  2.973348e+18  2016-05-07 22:52:05  ...  28.730140     Delhi       Delhi
1   2955066  4.734221e+18  2016-05-01 20:44:16  ...  22.660325  Calcutta  WestBengal
2    605968 -3.264500e+18  2016-05-02 14:23:04  ...  28.757906     Delhi       Delhi
3    448114  5.731369e+18  2016-05-03 13:21:16  ...  13.153332   Chennai   TamilNadu
4    665740  3.388880e+17  2016-05-06 03:51:05  ...  23.842609    Bokaro   Jharkhand

[5 rows x 7 columns]

In [6]:
events_data_df['state'].value_counts()

Delhi                       751733
Maharashtra                 677105
TamilNadu                   436282
AndhraPradesh               321332
MadhyaPradesh               232690
Rajasthan                   196361
WestBengal                  196161
Karnataka                    99058
UttarPradesh                 94750
Gujarat                      42881
Bihar                        35472
Punjab                       26195
Kerala                       23148
Haryana                      21053
Jharkhand                    18060
Orissa                       15904
Telangana                    12784
Assam                        12021
Chhattisgarh                  9754
Uttaranchal                   7720
JammuandKashmir               5385
Goa                           3250
Nagaland                      2298
Pondicherry                   1741
Mizoram                       1613
AndamanandNicobarIslands      1515
Meghalaya                     1515
HimachalPradesh               1409
Manipur             

Filtering out events_data for only 6 states as per the requirement given.

In [7]:
events_data_filtered_df= events_data_df[(events_data_df.state == 'WestBengal') | (events_data_df.state == 'Karnataka') | (events_data_df.state == 'Gujarat') | (events_data_df.state == 'Bihar') | (events_data_df.state == 'Punjab') | (events_data_df.state == 'Kerala')]

In [8]:
events_data_filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 422915 entries, 1 to 3252921
Data columns (total 7 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   event_id   422915 non-null  int64  
 1   device_id  422867 non-null  float64
 2   timestamp  422915 non-null  object 
 3   longitude  422873 non-null  float64
 4   latitude   422873 non-null  float64
 5   city       422915 non-null  object 
 6   state      422915 non-null  object 
dtypes: float64(3), int64(1), object(3)
memory usage: 25.8+ MB


Installing mysql connector for python

In [9]:
pip install mysql-connector-python-rf

     |████████████████████████████████| 11.9MB 270kB/s 
  Created wheel for mysql-connector-python-rf: filename=mysql_connector_python_rf-2.2.2-cp37-cp37m-linux_x86_64.whl size=249456 sha256=e7c5799e68e5abadb2754c1832ff0ed2ccebf7ba3236ee75e3929168aaaf12fe
  Stored in directory: /root/.cache/pip/wheels/87/58/fb/d95c84fad7e1bebfed324c13e107ebb08e1997c9226532859a
Successfully built mysql-connector-python-rf


import package to connect to mysql and loading data from tables to dataframes

In [10]:
import mysql.connector

#this is to connect to mysql database
mydb = mysql.connector.connect(
  host="cpanel.insaid.co",
  user="student",
  password="student",
  database="Capstone1"
)

#declaring a cursor variable of database to hold data of a specific table
mycursor = mydb.cursor()

#mycursor variable holds the entire data from gender_age_train table
mycursor.execute("SELECT * FROM gender_age_train")

#using fetchall function to get data from cursor and populating in another variable
myresult = mycursor.fetchall()

#inserting table data from variable into a dataframe
gender_age_train_df = pd.DataFrame(myresult,columns=['device_id','Gender','Age','Age_Group'])

#mycursor variable holds the entire data from phone_brand_device_model table
mycursor.execute("SELECT * FROM phone_brand_device_model")

#using fetchall function to get data from cursor and populating in another variable, this overwrites data from previous load
myresult = mycursor.fetchall()

#inserting table data from variable into a dataframe
phone_brand_device_model_df = pd.DataFrame(myresult,columns=['device_id','Brand','Model'])

#printing sample data from the dataframes
print("Data from Gender_Age_Train Table")
print("---------------------------------")
print(gender_age_train_df.head())
print("**************************************")
print("\n")
print("Data from Phone_Brand_Device_Model Table")
print("-----------------------------------------")
print(phone_brand_device_model_df.head())
print("**************************************")

Data from Gender_Age_Train Table
---------------------------------
             device_id Gender  Age Age_Group
0 -8076087639492063270      M   35    M32-38
1 -2897161552818060146      M   35    M32-38
2 -8260683887967679142      M   35    M32-38
3 -4938849341048082022      M   30    M29-31
4   245133531816851882      M   30    M29-31
**************************************


Data from Phone_Brand_Device_Model Table
-----------------------------------------
             device_id Brand    Model
0  1877775838486905855  vivo      Y13
1 -3766087376657242966    小米     V183
2 -6238937574958215831  OPPO      R7s
3  8973197758510677470    三星    A368t
4 -2015528097870762664    小米  红米Note2
**************************************


In [11]:
events_data_filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 422915 entries, 1 to 3252921
Data columns (total 7 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   event_id   422915 non-null  int64  
 1   device_id  422867 non-null  float64
 2   timestamp  422915 non-null  object 
 3   longitude  422873 non-null  float64
 4   latitude   422873 non-null  float64
 5   city       422915 non-null  object 
 6   state      422915 non-null  object 
dtypes: float64(3), int64(1), object(3)
memory usage: 25.8+ MB


In [12]:
gender_age_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74645 entries, 0 to 74644
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   device_id  74645 non-null  int64 
 1   Gender     74645 non-null  object
 2   Age        74645 non-null  int64 
 3   Age_Group  74645 non-null  object
dtypes: int64(2), object(2)
memory usage: 2.3+ MB


In [13]:
phone_brand_device_model_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87726 entries, 0 to 87725
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   device_id  87726 non-null  int64 
 1   Brand      87726 non-null  object
 2   Model      87726 non-null  object
dtypes: int64(1), object(2)
memory usage: 2.0+ MB


In [14]:
gender_age_train_df['Gender'].value_counts()

M    47904
F    26741
Name: Gender, dtype: int64

No entries in gender_age_train dataframe with null Gender

In [15]:
phone_brand_device_model_df['Brand'].value_counts(ascending=False)

小米      21645
三星      19211
华为      15181
vivo     7596
OPPO     6717
        ...  
瑞高          1
世纪星         1
恒宇丰         1
瑞米          1
E人E本        1
Name: Brand, Length: 116, dtype: int64

No records in phone_brand_device_model dataframe with null Brand

Merging Dataframes on device_id to create a single dataframe which contains customer age, gender, mobile brand and model

In [16]:
device_data_df=pd.merge(events_data_filtered_df,gender_age_train_df,on='device_id',how='inner') #performing inner join to fetch only matching device_id records
device_data_df.fillna(0,inplace=True) #replacing null values in the columns with 0
device_data_df

event_id     device_id            timestamp  ...  Gender  Age Age_Group
0      1836426  3.872522e+18  2016-05-01 20:29:14  ...       M   20      M22-
1       225132  3.872522e+18  2016-05-01 12:52:44  ...       M   20      M22-
2       815723  3.872522e+18  2016-05-01 12:28:56  ...       M   20      M22-
3      1300157  3.872522e+18  2016-05-04 23:24:57  ...       M   20      M22-
4       233682  3.872522e+18  2016-05-04 19:09:26  ...       M   20      M22-
...        ...           ...                  ...  ...     ...  ...       ...
2636   1530731 -2.288468e+17  2016-05-07 10:35:00  ...       F   63      F43+
2637   2967042 -2.288468e+17  2016-05-01 08:08:46  ...       F   63      F43+
2638   1334106 -2.734689e+18  2016-05-05 10:00:05  ...       M   27    M27-28
2639   1776773  3.711362e+18  2016-05-02 15:54:13  ...       M   31    M29-31
2640   1608869  3.501847e+18  2016-05-04 13:27:20  ...       M   28    M27-28

[2641 rows x 10 columns]

In [17]:
device_data_df=pd.merge(device_data_df,phone_brand_device_model_df,on='device_id',how='inner')
device_data_df.fillna(0,inplace=True)
device_data_df

event_id     device_id            timestamp  ...  Age_Group  Brand      Model
0      1836426  3.872522e+18  2016-05-01 20:29:14  ...       M22-     酷派       8702
1       225132  3.872522e+18  2016-05-01 12:52:44  ...       M22-     酷派       8702
2       815723  3.872522e+18  2016-05-01 12:28:56  ...       M22-     酷派       8702
3      1300157  3.872522e+18  2016-05-04 23:24:57  ...       M22-     酷派       8702
4       233682  3.872522e+18  2016-05-04 19:09:26  ...       M22-     酷派       8702
...        ...           ...                  ...  ...        ...    ...        ...
2636   1530731 -2.288468e+17  2016-05-07 10:35:00  ...       F43+     魅族        MX3
2637   2967042 -2.288468e+17  2016-05-01 08:08:46  ...       F43+     魅族        MX3
2638   1334106 -2.734689e+18  2016-05-05 10:00:05  ...     M27-28     华为    荣耀3X畅玩版
2639   1776773  3.711362e+18  2016-05-02 15:54:13  ...     M29-31     三星  Galaxy S6
2640   1608869  3.501847e+18  2016-05-04 13:27:20  ...     M27-28     小米         E5

[2641 rows x 12 columns]

In [18]:
events_data_filtered_df['device_id'].nunique()

19032

For the 6 states in consideration there are 19032 devices.

In [19]:
device_data_df['device_id'].nunique()

120

For all the 6 states under consideration we've customer data for only 120 devices.

In [20]:
((events_data_filtered_df['device_id'].nunique() - device_data_df['device_id'].nunique()) / events_data_filtered_df['device_id'].nunique()) * 100

99.36948297604036

1.   Only 2641 records have matching device_id from all the 3 dataframes.
2.   Of 2641 matching records there are 120 unique DeviceIDs which mean that there are multiple events for few Devices.
3.    For all the 6 states under consideration customer data is not available for ~99.4% of the devices.


In [21]:
device_data_df['Model'].nunique()

85

In [22]:
device_data_df['Brand'].nunique()

14

Upon merging all the 3 dataframes we've data for around 85 models from 14 Brands.

In [23]:
device_data_df.head()

event_id     device_id            timestamp  ...  Age_Group  Brand Model
0   1836426  3.872522e+18  2016-05-01 20:29:14  ...       M22-     酷派  8702
1    225132  3.872522e+18  2016-05-01 12:52:44  ...       M22-     酷派  8702
2    815723  3.872522e+18  2016-05-01 12:28:56  ...       M22-     酷派  8702
3   1300157  3.872522e+18  2016-05-04 23:24:57  ...       M22-     酷派  8702
4    233682  3.872522e+18  2016-05-04 19:09:26  ...       M22-     酷派  8702

[5 rows x 12 columns]

In [24]:
((device_data_df[device_data_df['Gender'] == 'M']['Gender'].count() - device_data_df[device_data_df['Gender'] == 'F']['Gender'].count()) / device_data_df[device_data_df['Gender'] == 'F']['Gender'].count()) * 100

73.67875647668394

73.7% of the customers are Male.

In [25]:
device_data_df['Age_Group'].value_counts()

M23-26    604
M29-31    330
F27-28    325
M32-38    322
M22-      289
F29-32    234
F23-      153
F43+      114
M39+       80
F33-42     75
F24-26     64
M27-28     51
Name: Age_Group, dtype: int64

In [26]:
device_data_df['state'].value_counts()

WestBengal    959
Gujarat       707
Karnataka     572
Bihar         300
Kerala         71
Punjab         32
Name: state, dtype: int64

In [27]:
((device_data_df['state'].count() - device_data_df[device_data_df['state'] == 'WestBengal']['state'].count()) / device_data_df['state'].count()) * 100

63.687996970844374

In [28]:
device_data_df[device_data_df['state'] == 'WestBengal']['city'].value_counts()

Calcutta                787
Durgapur                 50
Kharagpur                43
Ghatal                   28
Kandi                    11
Kalna                     8
Pujali                    6
Puruliya                  6
Contai                    6
Chandannagar              4
Chakdaha                  3
Naihati                   2
Ulubaria                  2
AshoknagarKalyangarh      2
Shiliguri                 1
Name: city, dtype: int64

In [29]:
device_data_df['city'].value_counts()

Calcutta                787
Bangalore               327
Bardoli                 285
Porbandar               224
Saharsa                 151
DodBallapur              75
Amreli                   69
Sira                     55
Jetpur                   50
Durgapur                 50
Kozhikkod                49
Yelahanka                43
Kharagpur                43
Bhagalpur                40
Supaul                   37
Siddhapur                32
Puttur                   31
Ghatal                   28
Surat                    21
Munger                   20
Tumkur                   19
Phulwari                 18
Dhoraji                  16
Ponnani                  14
Sirsi                    14
Amritsar                 11
Kandi                    11
Kishanganj                9
Thrissur                  8
Sitamarhi                 8
Kalna                     8
Jahanabad                 7
Challakere                7
Pujali                    6
Puruliya                  6
Contai              

In [30]:
((device_data_df[device_data_df['state'] == 'WestBengal']['state'].count() - device_data_df[device_data_df['city'] == 'Calcutta']['city'].count()) / device_data_df[device_data_df['state'] == 'WestBengal']['state'].count()) * 100

17.935349322210637

In [31]:
((device_data_df['device_id'].count() - device_data_df[device_data_df['city'] == 'Calcutta']['city'].count()) / device_data_df['device_id'].count()) * 100

70.20068156001514

In [32]:
((device_data_df['device_id'].count() - device_data_df[device_data_df['city'] == 'Bangalore']['city'].count()) / device_data_df['device_id'].count()) * 100

87.61832639151837

In [34]:
device_data_df.head()

event_id     device_id            timestamp  ...  Age_Group  Brand Model
0   1836426  3.872522e+18  2016-05-01 20:29:14  ...       M22-     酷派  8702
1    225132  3.872522e+18  2016-05-01 12:52:44  ...       M22-     酷派  8702
2    815723  3.872522e+18  2016-05-01 12:28:56  ...       M22-     酷派  8702
3   1300157  3.872522e+18  2016-05-04 23:24:57  ...       M22-     酷派  8702
4    233682  3.872522e+18  2016-05-04 19:09:26  ...       M22-     酷派  8702

[5 rows x 12 columns]

In [36]:
device_data_df['Brand'].value_counts(ascending=False)

小米      941
三星      577
华为      369
vivo    205
酷派      161
OPPO     95
魅族       94
乐视       72
HTC      57
一加       35
亿通       17
努比亚       9
酷比魔方      8
锤子        1
Name: Brand, dtype: int64

In [40]:
((device_data_df['device_id'].count() - device_data_df[device_data_df['Brand']=='小米']['device_id'].count()) / device_data_df['device_id'].count()) * 100

64.36955698599014

In [42]:
device_data_df['Model'].value_counts(ascending=False)

红米note        306
红米Note2       224
Galaxy Win    151
MI 3          137
荣耀畅玩4         115
             ... 
红米note增强版       2
T1              1
E5              1
Z9              1
超级手机1           1
Name: Model, Length: 85, dtype: int64

In [46]:
device_data_df['Model'].value_counts(ascending=False)

2641

In [44]:
device_data_df[device_data_df['Model']=='红米note']

event_id     device_id            timestamp  ...  Age_Group  Brand   Model
1056    271076  1.274328e+17  2016-05-03 07:37:31  ...     F27-28     小米  红米note
1057    251055  1.274328e+17  2016-05-03 12:32:48  ...     F27-28     小米  红米note
1058   2787201  1.274328e+17  2016-05-07 19:38:26  ...     F27-28     小米  红米note
1059   3056922  1.274328e+17  2016-05-03 08:19:13  ...     F27-28     小米  红米note
1060   2177734  1.274328e+17  2016-05-01 08:37:27  ...     F27-28     小米  红米note
...        ...           ...                  ...  ...        ...    ...     ...
1738    201154  6.737477e+16  2016-05-03 12:33:27  ...     F24-26     小米  红米note
1739    792606  6.737477e+16  2016-05-07 12:29:59  ...     F24-26     小米  红米note
1740    201940  6.737477e+16  2016-05-07 12:29:47  ...     F24-26     小米  红米note
2631   1652210  5.939461e+16  2016-05-05 01:18:22  ...     F33-42     小米  红米note
2632     30264  5.939461e+16  2016-05-01 01:18:27  ...     F33-42     小米  红米note

[306 rows x 12 columns]

In [45]:
((device_data_df['Model'].count() - device_data_df[device_data_df['Model']=='红米note']['device_id'].count()) / device_data_df['Model'].count()) * 100

88.41347974252177

# Observations


1.   There are 19032 devices with event data for the 6 states being considered of which customer data w.r.t mobile model, brand, customer details are present for ~0.6% only.
2.   Among the 0.6% of Customer data available there are 85 Models of mobiles from among 14 Brands are being used.
3. Among the 0.6% of customer data available for the 6 States.
4. Among the 0.6% of available customer data 73.7% of them are Male.
5. Most of the Male customers are in the age group of 23-26 & 29-31.
6. Most of the Female customers are in the age group of 27-28 followed by 29-32.
7. Of the available customer data most (~36.4%) of them are from WestBengal followed by Gujarat, Karnataka & Bihar states.
8. 82% of the customers in WestBengal state are from Calcutta.
9. Calcutta constitutes to ~29.8% of customers followed by Bangalore which constitutes to ~12.4%.
10. Of the device data available ~35.6% of them belong to the brand Xiaomi.
11. Of the device data available ~11.6% of them belong to Note model of Xiaomi brand.


